Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -498,10 +498,12 @@ class BrowserViewController: UIViewController {
         let alert = UIAlertController.restoreTabsAlert(
             okayCallback: { _ in
                 self.tabManager.restoreTabs()
+                self.updateTabCountUsingTabManager(self.tabManager, animated: true)
```

Comment:
Any reason why `restoreTabs()` doesn't trigger a tab count update? This seems very manual / error prone. I know `updateTabCountUsingTabManager` is in different class, but is there some tab creation hook we can leverage?

Reply:
It really _should_ be doing it after adding the new tabs, but then fails to do so because it marks itself as "restoring" so the code around doing it fails and the logic around why they have that seems more fragile to remove than to do this.

```swift
if !tabManager.isRestoring {
    updateTabCountUsingTabManager(tabManager)
}
```

What's more confusing is they prefix these lines of code with `// If we are restoring tabs then we update the count once at the end` but actually doesn't restore.

Then when it _can_ count it, theres no selected tab yet because the logic is "addTab → selectTab" and selecting a tab doesn't update the tab count because there was no previous tab. That logic is this:

```swift
if let selected = selected, let previous = previous, selected.isPrivate != previous.isPrivate {
    updateTabCountUsingTabManager(tabManager)
}
```

So basically it never updates... But the non-restoring flow calls the `self.updateTabCountUsingTabManager` method.

Reply:
Something is either broken with the delegate mechanism or some weird timing condition regarding when the delegates are notified:

`tabManager.addDelegate(self)`

```    
func tabManagerDidRestoreTabs(_ tabManager: TabManager) {
     updateTabCountUsingTabManager(tabManager)
}
```

This feels like we're just throwing a bandaid on this, but manually remembering to add tab updates everywhere _will_ break at some point. I'd rather try to fix the setup that is already in place. Either we broke it or inherited some regression.

Reply:
OK I'll give it another pass and figure out what's going on here, removing [needsuplift] for now